pkg/internal/appolly/appolly.go

```diff
@@ -44,45 +44,37 @@ func New(ctx context.Context, ctxInfo *global.ContextInfo, config *beyla.Config)
 
 // FindAndInstrument searches in background for any new executable matching the
 // selection criteria.
-func (i *Instrumenter) FindAndInstrument(wg *sync.WaitGroup) error {
+// Returns a channel that is closed when the Instrumenter completed all its tasks.
+// This is: when the context is cancelled, it has unloaded all the eBPF probes.
+func (i *Instrumenter) FindAndInstrument() (<-chan struct{}, error) {
 	finder := discover.NewProcessFinder(i.ctx, i.config, i.ctxInfo, i.tracesInput)
 	foundProcesses, deletedProcesses, err := finder.Start()
 	if err != nil {
-		return fmt.Errorf("couldn't start Process Finder: %w", err)
+		return nil, fmt.Errorf("couldn't start Process Finder: %w", err)
 	}
+
+	done := make(chan struct{})
 	// In background, listen indefinitely for each new process and run its
 	// associated ebpf.ProcessTracer once it is found.
-	wg.Add(1)
+	wg := sync.WaitGroup{}
 	go func() {
-		defer wg.Done()
 		log := log()
-		type cancelCtx struct {
-			ctx    context.Context
-			cancel func()
-		}
-		contexts := map[uint64]cancelCtx{}
```

Comment:
Didn't saw the utility of storing all the contexts and explicitly cancelling them after the parent context is cancelled.

If the parent context its Done, all the child contexts are Done in cascade.